packages/snaps-controllers/src/interface/utils.ts

```diff
@@ -1,102 +1,185 @@
-import { NodeType, assert } from '@metamask/snaps-sdk';
+import { assert } from '@metamask/snaps-sdk';
 import type {
-  Component,
-  Input,
   FormState,
   InterfaceState,
+  ComponentOrElement,
 } from '@metamask/snaps-sdk';
+import type {
+  ButtonElement,
+  FieldElement,
+  InputElement,
+  JSXElement,
+} from '@metamask/snaps-sdk/jsx';
+import { isJSXElementUnsafe } from '@metamask/snaps-sdk/jsx';
+import { getJsxElementFromComponent } from '@metamask/snaps-utils';
 
 /**
- * Construct the state for a stray input (not enclosed in a form).
+ * Get a JSX element from a component or JSX element. If the component is a
+ * JSX element, it is returned as is. Otherwise, the component is converted to
+ * a JSX element.
  *
- * @param state - The interface state.
- * @param component - The Input component.
+ * @param component - The component to convert.
+ * @returns The JSX element.
+ */
+export function getJsxInterface(component: ComponentOrElement): JSXElement {
+  if (isJSXElementUnsafe(component)) {
+    return component;
+  }
+
+  return getJsxElementFromComponent(component);
+}
+
+/**
+ * Assert that the component name is unique in state.
+ *
+ * @param state - The interface state to verify against.
+ * @param name - The component name to verify.
+ */
+export function assertNameIsUnique(state: InterfaceState, name: string) {
+  assert(
+    state[name] === undefined,
+    `Duplicate component names are not allowed, found multiple instances of: "${name}".`,
+  );
+}
+
+/**
+ * Construct the state for an input field.
+ *
+ * @param oldState - The previous state.
+ * @param element - The input element.
  * @returns The input state.
  */
-export const constructInputState = (
-  state: InterfaceState,
-  component: Input,
-) => {
-  return component.value ?? state[component.name] ?? null;
-};
+function constructJsxInputState(
+  oldState: InterfaceState,
+  element: InputElement,
+) {
+  return element.props.value ?? oldState[element.props.name] ?? null;
+}
 
 /**
  * Construct the state for a form input.
  *
- * Sets the state to either the specified component value, the previous value from the old state or null.
- *
- * @param state - The interface state.
- * @param component - The Input component.
+ * @param oldState - The previous state.
+ * @param component - The input element.
  * @param form - The parent form name of the input.
  * @returns The input state.
  */
-export const constructFormInputState = (
-  state: InterfaceState,
-  component: Input,
+function constructJsxFormInputState(
```

Comment:
Do we need the name changes in this file? If everything is JSX, this feels excessive 😅

Reply:
It was a separate function initially but I removed the legacy implementation later. I can rename it back if you prefer.

Reply:
Yes please!